perceval/converters/myqlm_converter.py

```diff
@@ -0,0 +1,151 @@
+# copies or substantial portions of the Software.
+#
+# As a special exception, the copyright holders of exqalibur library give you
+# permission to combine exqalibur with code included in the standard release of
+# Perceval under the MIT license (or modified versions of such code). You may
+# copy and distribute such a combined system following the terms of the MIT
+# license for both exqalibur and Perceval. This exception for the usage of
+# exqalibur is limited to the python bindings used by Perceval.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from perceval.components import Port, Circuit, Processor, Source, BS, PS
+from perceval.utils import P, BasicState, Encoding
+from perceval.utils.algorithms.optimize import optimize
+from perceval.utils.algorithms.norm import frobenius
+import perceval.components.unitary_components as comp
+import numpy as np
+
+
+min_precision_gate = 1e-4
+
+
+class MyQLMConverter:
+    r"""myQLM quantum circuit to perceval circuit converter.
+    :param catalog: component library of perceval
+    """
+    def __init__(self, catalog, backend_name: str = "SLOS", source: Source = Source()):
+        self._source = source
+        self._backend_name = backend_name
+        self._heralded_cnot_builder = catalog["heralded cnot"]
+        self._heralded_cz_builder = catalog["heralded cz"]
+        self._postprocessed_cnot_builder = catalog["postprocessed cnot"]
+        self._generic_2mode_builder = catalog["generic 2 mode circuit"]
+        self._lower_phase_component = Circuit(2) // (0, comp.PS(P("phi2")))
+        self._upper_phase_component = Circuit(2) // (1, comp.PS(P("phi1")))
+        self._two_phase_component = Circuit(2) // (0, comp.PS(P("phi1"))) // (1, comp.PS(P("phi2")))
+
+    def convert(self, qlmc, use_postselection: bool = True) -> Processor:
+        r"""Convert a myQLM quantum circuit into a perceval `Processor`.
+
+        :param qlmc: quantum gate-based myqlm circuit
+        :type qlmc: qat.core.Circuit
+        :param use_postselection: when True, uses a `post-processed CNOT` as the last gate. Otherwise, uses only
+            `heralded CNOT`
+        :return: the converted Processor
+        """
+        import qat
+        from qat.core.circuit_builder.matrix_util import circ_to_np
+        # importing the quantum toolbox of myqlm
+        # this nested import fixes automatic class reference generation
+
+        # count the number of CNOT gates to use during the conversion, will give us the number of herald to handle
+        n_cnot = qlmc.count("CNOT")
```

Comment:
Consider declaring variable only when you need it, and so moving `n_cnot` and `cnot_idx` line 82 after the first for loop